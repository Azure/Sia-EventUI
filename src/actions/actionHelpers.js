import { authenticatedFetch, authenticatedPost, authenticatedPut } from 'services/authenticatedFetch'

export const testableReduxBackedPromise = (localAuthenticatedFetch, localAuthenticatedPost, localAuthenticatedPut) =>
(promiseArgs, actionSet, operation = 'GET') =>
(dispatch) => {
  validateActionSet(actionSet)

  const promiseGenerator = getPromiseGenerator(localAuthenticatedFetch, localAuthenticatedPost, localAuthenticatedPut)(operation)

  dispatch(actionSet.try())

  return promiseGenerator(dispatch, ...promiseArgs)
        .then(({json, response}) => dispatch(actionSet.succeed(json, response)),
            error => dispatch(actionSet.fail(error)))
}

export const reduxBackedPromise = testableReduxBackedPromise(authenticatedFetch, authenticatedPost, authenticatedPut)

const needOnActionSet = (prop) => `Need "${prop}" function on actionSet!`

export const validateActionSet = (actionSet) => {
  if (!actionSet.try || !(typeof actionSet.try === 'function')) { throw needOnActionSet('try') }
  if (!actionSet.succeed || !(typeof actionSet.succeed === 'function')) { throw needOnActionSet('succeed') }
  if (!actionSet.fail || !(typeof actionSet.fail === 'function')) { throw needOnActionSet('fail') }
}

export const getPromiseGenerator = (localAuthenticatedFetch, localAuthenticatedPost, localAuthenticatedPut) =>
(operation) => {
  switch (operation.toUpperCase()) {
    case 'TESTERROR':
      break // Never intended to happen in a deployed instance, just here for testing
    case 'PUT':
      return localAuthenticatedPut
    case 'POST':
      return localAuthenticatedPost
    default:
      return localAuthenticatedFetch
  }
  throw new Error('promiseGenerator not initialized. This should not be possible. Consider rolling back.')
}

const goToPageActionType = (BASE_NAME) => 'GOTO_' + BASE_NAME + '_PAGE'
const nextPageActionType = (BASE_NAME) => 'NEXT_' + BASE_NAME + '_PAGE'
const prevPageActionType = (BASE_NAME) => 'PREV_' + BASE_NAME + '_PAGE'
const sortActionType = (BASE_NAME) => 'SORT_' + BASE_NAME
const filterActionType = (BASE_NAME) => 'FILTER_' + BASE_NAME

export const paginationActions = (BASE_NAME) => ({
  types: {
    GOTO_PAGE: goToPageActionType(BASE_NAME),
    NEXT_PAGE: nextPageActionType(BASE_NAME),
    PREV_PAGE: prevPageActionType(BASE_NAME),
    SORT: sortActionType(BASE_NAME),
    FILTER: filterActionType(BASE_NAME)
  },

  goToPage: (page) => ({
    type: goToPageActionType(BASE_NAME),
    page
  }),

  nextPage: () => ({
    type: nextPageActionType(BASE_NAME)
  }),

  sort: (by) => ({
    type: sortActionType(BASE_NAME),
    by
  }),

  filter: (filter) => ({
    type: filterActionType(BASE_NAME),
    filter
  })
})

/* Update pagination does nothing on its own, but activates
the default case of pagination reducers created by paginated-redux,
which forces re-evaluation of the current page; simply adding
items to the list array won't cause the contents of the current page
to change.
*/
export const UPDATE_PAGINATION = 'UPDATE_PAGINATION'
export const updatePagination = () => ({type: UPDATE_PAGINATION})
