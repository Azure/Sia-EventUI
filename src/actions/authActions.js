export const LOGIN_IN_PROGRESS = 'LOGIN_IN_PROGRESS'
export const USER_LOGGED_IN = 'USER_LOGGED_IN'
export const USER_LOGGED_OUT = 'USER_LOGGED_OUT'
export const USER_LOGIN_ERROR = 'USER_LOGIN_ERROR'

export const loginInProgress = () => ({
  type: LOGIN_IN_PROGRESS
})

export const userLoggedIn = (user) => ({
  type: USER_LOGGED_IN,
  user
})

export const userLoggedOut = () => ({
  type: USER_LOGGED_OUT
})

export const userLoginError = error => ({
  type: USER_LOGIN_ERROR,
  error
})
