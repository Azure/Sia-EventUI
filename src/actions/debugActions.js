/*
These actions have no associated reducer, but they provide additional context for debugging
without changing state directly. They're intended to be used with Redux dev tools.
*/

const RAW_HTTP_RESPONSE = 'DEBUG_RAW_HTTP_RESPONSE'
const JSON_RESULT = 'DEBUG_JSON_RESULT'

// response is not serializable
export const rawHttpResponse = (response) => ({
  type: RAW_HTTP_RESPONSE,
  response: response ? {
    bodyUsed: response.bodyUsed,
    ok: response.ok,
    redirected: response.redirected,
    status: response.status,
    statusText: response.statusText,
    type: response.type,
    url: response.url
  } : null
})

export const jsonResult = (json) => ({
  type: JSON_RESULT,
  json
})
