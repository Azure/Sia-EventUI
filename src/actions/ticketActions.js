export const UPDATE_TICKET_QUERY = 'UPDATE_TICKET_QUERY'
export const REMOVE_TICKET = 'REMOVE_TICKET'
export const REMOVE_PREVIOUS_TICKETS = 'REMOVE_PREVIOUS_TICKETS'

export const updateTicketQuery = ticketQuery => ({
  type: UPDATE_TICKET_QUERY,
  ticketQuery
})

export const removePreviousTicketsFromRecent = (currentTicketId) => ({
  type: REMOVE_PREVIOUS_TICKETS,
  current: currentTicketId
})

export const removeTicketFromRecent = (id) => ({
  type: REMOVE_TICKET,
  id
})
