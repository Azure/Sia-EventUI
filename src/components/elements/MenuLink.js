import React from 'react'
import { Link } from 'react-router-dom'
import MenuItem from 'material-ui/MenuItem'
import ActionDelete from 'material-ui/svg-icons/action/delete'

export const MenuLink = (type, id, onClick, dispatch) => {
  var typeTitle = makeTitle(type)
  var typeRoute = makeRoute(type, id)
  return (
    <MenuItem
      key={makeKey(type, id)}
      primaryText={<Link to={typeRoute} >{`${typeTitle} ${id}`}</Link>}
      rightIcon={<ActionDelete onClick={ () => dispatch(onClick(id)) } />}
    />
  )
}

const menuItemOnClick = (onClick, dispatch, id) => {
  dispatch(onClick(id))
}

const makeTitle = (type) => {
  return type.charAt(0).toUpperCase() + type.slice(1)
}

const makeRoute = (type, id) => {
  return `/${ type }s/${ id }`
}

const makeKey = (type, id) => {
  return `${type}-${id}`
}

export default MenuLink
