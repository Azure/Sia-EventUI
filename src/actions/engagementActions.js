import { DateTime } from 'luxon'
import { reduxBackedPromise, paginationActions } from 'actions/actionHelpers'

export const TRY_ENGAGE = 'TRY_ENGAGE'
export const ENGAGE_SUCCESS = 'ENGAGE_SUCCESS'
export const ENGAGE_FAILURE = 'ENGAGE_FAILURE'
export const TRY_DISENGAGE = 'TRY_DISENGAGE'
export const DISENGAGE_SUCCESS = 'DISENGAGE_SUCCESS'
export const DISENGAGE_FAILURE = 'DISENGAGE_FAILURE'
export const ENGAGEMENTS = 'ENGAGEMENTS'

export const engage = (incidentId, participant, timeEngaged = DateTime.utc()) =>
reduxBackedPromise(
    ['incidents/' + incidentId + '/engagements/', {participant}],
    engageActionSet(incidentId, participant, timeEngaged),
    'POST'
)

export const disengage = (participant, engagement, timeDisengaged = DateTime.utc()) =>
reduxBackedPromise(
  [
    'incidents/' + engagement.incidentId + '/engagements/' + engagement.id,
    updatedEngagement(engagement, timeDisengaged),
    null,
    false
  ],
    disengageActionSet(engagement.incidentId, participant, engagement, timeDisengaged),
    'PUT'
)

const engageActionSet = (incidentId, participant, timeEngaged) => ({
  try: () => ({
    type: TRY_ENGAGE,
    incidentId,
    participant,
    timeEngaged
  }),

  succeed: (engagement) => ({
    type: ENGAGE_SUCCESS,
    engagement
  }),

  fail: (error) => ({
    type: ENGAGE_FAILURE,
    error,
    participant,
    timeEngaged
  })
})

const updatedEngagement = (engagement, timeDisengaged) => ({
  ...engagement,
  timeDisengaged
})

const disengageActionSet = (incidentId, participant, engagement, timeDisengaged) => ({
  try: () => ({
    type: TRY_DISENGAGE,
    incidentId,
    participant,
    timeDisengaged
  }),

  succeed: () => ({
    type: DISENGAGE_SUCCESS,
    engagement: updatedEngagement(engagement, timeDisengaged)
  }),

  fail: (error) => ({
    type: DISENGAGE_FAILURE,
    error,
    participant,
    timeDisengaged
  })
})

export const pagination = paginationActions(ENGAGEMENTS)
