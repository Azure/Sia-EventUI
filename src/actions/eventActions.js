import moment from 'moment'
import { paginationActions, updatePagination, reduxBackedPromise } from 'actions/actionHelpers'
import * as filterActions from 'actions/filterActions'

export const EVENTS = 'EVENTS'
export const REQUEST_EVENT = 'REQUEST_EVENT'
export const RECEIVE_EVENT = 'RECEIVE_EVENT'
export const RECEIVE_EVENT_FAILURE = 'RECEIVE_EVENT_FAILURE'
export const REQUEST_EVENTS = 'REQUEST_EVENTS'
export const RECEIVE_EVENTS = 'RECEIVE_EVENTS'
export const RECEIVE_EVENTS_FAILURE = 'RECEIVE_EVENTS_FAILURE'
export const POST_EVENT_TRY = 'POST_EVENT_TRY'
export const POST_EVENT_SUCCEED = 'POST_EVENT_SUCCEED'
export const POST_EVENT_FAIL = 'POST_EVENT_FAIL'
export const ADD_EVENT = 'ADD_EVENT'

export const pagination = paginationActions(EVENTS)

export const linksHeaderName = 'links'

export const fetchEvent = (incidentId, eventId) => reduxBackedPromise(
    [getEventsEndPoint(incidentId) + eventId],
    getEventActionSet(incidentId, eventId)
)

export const fetchEvents = (filter) => reduxBackedPromise(
        getEventsFetchArgs(filter),
        getEventsActionSet(filter.incidentId)
    )

export const postEvent = (incidentId, eventTypeId = 0, data = {}, occurrenceTime = moment()) => reduxBackedPromise(
    postEventFetchArgs(incidentId, eventTypeId, data, occurrenceTime),
    postEventActionSet(incidentId),
    'POST'
)

export const getEventsEndPoint = (incidentId) => (incidentId ? 'incidents/' + incidentId + '/' : '') + 'events/'

export const getEventsFetchArgs = (filter) => ([
  getEventsEndPoint(filter.incidentId) + filterActions.serializeFiltersForUrl(filter)
])

export const getEventFetchArgs = (incidentId, eventId) => {
  return [getEventsEndPoint(incidentId) + eventId]
}

export const postEventFetchArgs = (incidentId, eventTypeId, data, occurrenceTime) => ([
  getEventsEndPoint(incidentId),
  {
    eventTypeId,
    occurred: occurrenceTime,
    eventFired: occurrenceTime,
    data
  }
])

export const getEventActionSet = (incidentId, eventId) => ({
  try: () => ({
    type: REQUEST_EVENT,
    incidentId,
    id: eventId
  }),

  succeed: (event) => (dispatch) => {
    dispatch({
      type: RECEIVE_EVENT,
      event,
      id: eventId
    })

    dispatch(updatePagination())
  },

  fail: (failureReason) => ({
    type: RECEIVE_EVENT_FAILURE,
    failureReason,
    incidentId,
    id: eventId
  })
})

export const getEventsActionSet = (incidentId) => ({
  try: () => ({
    type: REQUEST_EVENTS,
    incidentId
  }),

  succeed: (events, response) => (dispatch) => {
    let linksHeader
    for (let header of response.headers) {
      if (header[0] === linksHeaderName) {
        linksHeader = JSON.parse(header[1])
      }
    }

    dispatch({
      type: RECEIVE_EVENTS,
      events,
      incidentId,
      pagination: linksHeader
    })

    if (linksHeader.NextPageLink) {
      dispatch(reduxBackedPromise(
                [linksHeader.NextPageLink],
                getEventsActionSet(incidentId)
            ))
    } else {
      dispatch(updatePagination())
    }
  },

  fail: (failureReason) => ({
    type: RECEIVE_EVENTS_FAILURE,
    failureReason,
    incidentId
  })
})

export const addEvent = (event, incidentId) => ({
  type: ADD_EVENT,
  incidentId,
  event
})

export const postEventActionSet = (incidentId) => ({
  try: () => ({
    type: POST_EVENT_TRY,
    incidentId
  }),

  succeed: (event) => dispatch => {
    dispatch({
      type: POST_EVENT_SUCCEED,
      incidentId,
      event
    })

    dispatch(updatePagination())
  },

  fail: (failureReason) => ({
    type: POST_EVENT_FAIL,
    incidentId,
    failureReason
  })
})
