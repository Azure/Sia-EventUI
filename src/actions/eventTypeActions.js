import { reduxBackedPromise } from 'actions/actionHelpers'

export const TRY_GET_EVENT_TYPE = 'TRY_GET_EVENT_TYPE'
export const GET_EVENT_TYPE_SUCCESS = 'GET_EVENT_TYPE_SUCCESS'
export const GET_EVENT_TYPE_FAILURE = 'GET_EVENT_TYPE_FAILURE'

export const TRY_GET_EVENT_TYPES = 'TRY_GET_EVENT_TYPES'
export const GET_EVENT_TYPES_SUCCESS = 'GET_EVENT_TYPES_SUCCESS'
export const GET_EVENT_TYPES_FAILURE = 'GET_EVENT_TYPES_FAILURE'

export const fetchEventType = (eventTypeId) => reduxBackedPromise(
    ['eventTypes/' + eventTypeId],
    getEventTypeActionSet(eventTypeId)
)

export const fetchEventTypes = () => reduxBackedPromise(
    ['eventTypes/'], getEventTypesActionSet
)

export const getEventTypeActionSet = (eventTypeId) => ({
  try: () => ({
    type: TRY_GET_EVENT_TYPE,
    id: eventTypeId
  }),

  succeed: (eventType) => ({
    type: GET_EVENT_TYPE_SUCCESS,
    id: eventTypeId,
    eventType
  }),

  fail: (failureReason) => ({
    type: GET_EVENT_TYPE_FAILURE,
    id: eventTypeId,
    failureReason
  })
})

export const getEventTypesActionSet = ({
  try: () => ({
    type: TRY_GET_EVENT_TYPES
  }),

  succeed: (eventTypes) => ({
    type: GET_EVENT_TYPES_SUCCESS,
    eventTypes
  }),

  fail: (failureReason) => ({
    type: GET_EVENT_TYPES_FAILURE,
    failureReason
  })
})
