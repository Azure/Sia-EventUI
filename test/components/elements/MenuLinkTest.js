'use strict'
import { expect } from 'chai'
import React from 'react'
import { Link } from 'react-router-dom'
import MenuItem from 'material-ui/MenuItem'
import ActionDelete from 'material-ui/svg-icons/action/delete'
import { removeTicketFromRecent } from 'actions/ticketActions'
import MenuLink, { deleteMenuItem } from 'components/elements/MenuLink'

describe('MenuLink', function() {
  describe('when all input is valid', function() {
    const mockType = 'ticket'
    const mockId = 0
    const mockOnClick = () => 'clicked'
    const mockDispatch = (input) => input

    const result = MenuLink(mockType, mockId, mockOnClick, mockDispatch)

    it('should render a MenuItem', function() {
      expect(result.type).to.equal(MenuItem)
    })

    it('should contain a Link in the primaryText', function() {
      expect(result.props.primaryText.type).to.equal(Link)
    })

    it('should link to a route composed of type and id', function() {
      expect(result.props.primaryText.props.to).to.equal(`/${mockType}s/${mockId}`)
    })

    it('should have a route named for the type and id', function () {
      const mockLinkText = mockType.charAt(0).toUpperCase() + mockType.slice(1)
      expect(result.props.primaryText.props.children).to.equal(`${mockLinkText} ${mockId}`)
    })

    it('should have a key that is composed of both the type and id', function() {
      expect(result.key).to.equal(`${mockType}-${mockId}`)
    })

    it('should contain a rightIcon of type ActionDelete', function() {
      expect(result.props.rightIcon.type).to.equal(ActionDelete)
    })

    it('should contain a rightIcon that sets off a function', function() {
      expect(typeof(result.props.rightIcon.props.onClick)).to.equal('function')
    })
  })

  describe('when type input is not valid', function() {
    const mockId = 0
    const mockOnClick = () => 'clicked'
    const mockDispatch = (input) => input

    describe('when type is null', function() {
      const mockType = null
      it('should throw an error', function() {
        expect(() => MenuLink(mockType, mockId, mockOnClick, mockDispatch)).to.throw('the type input is invalid')
      })
    })

    describe('when type is undefined', function () {
      const mockType = undefined
      it('should throw an error', function () {
        expect(() => MenuLink(mockType, mockId, mockOnClick, mockDispatch)).to.throw('the type input is invalid')
      })
    })

    describe('when type input is less than 2 characters', function () {
      const mockType = ''
      it('should throw an error', function () {
        expect(() => MenuLink(mockType, mockId, mockOnClick, mockDispatch)).to.throw('the type input is invalid')
      })
    })
  })
})
