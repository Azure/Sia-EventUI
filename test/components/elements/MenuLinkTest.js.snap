// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuLink when all input is valid should match the snapshot 1`] = `
<MenuItem
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  checked={false}
  containerElement={
    <Link
      replace={false}
      to="/mock/route"
    />
  }
  desktop={false}
  disabled={false}
  focusState="none"
  insetChildren={false}
  key="mock title/mock/route"
  onClick={[Function]}
  primaryText={
    <div
      style={
        Object {
          "overflow": "hidden",
          "textOverflow": "ellipsis",
          "width": "25rem",
        }
      }
    >
      mock title
    </div>
  }
  rightIcon={Object {}}
  targetOrigin={
    Object {
      "horizontal": "left",
      "vertical": "top",
    }
  }
  title="mock title"
/>
`;
